### ISSUE TITLE
✨ [Typescript] eslint rule to avoid some common gotchas

### ISSUE THREAD

**Issue by Xiot** [dagger/dagger#7402]
### What are you trying to do?

While trying to get around https://github.com/dagger/dagger/issues/7401, I tried to keep the `@field` as a string, but then just type a custom property.

```typescript
@object()
export class MergeRequestReference {
  @field() _iid: string
  @field() _type: string

  constructor(iid: string, type: 'head' | 'merge') {
    this._iid = iid
    this._type = type
  }

  get iid() {
    return this._iid
  }
  get type() {
    return this._type
  }
}
```

However it seems that the prototype isn't actually preserved when the object is used from another module. 

First prize, would be to re-attach the prototype chain, but barring that, it would be nice to have an eslint rule to catch things that work slightly differently than a typescript developer would expect.

### Why is this important to you?

It would avoid some errors caused by the way that dagger modules are executed.

### How are you currently working around this?

I just removed the custom getters, and I check the types later in the code

**Comment by TomChv** [comment]
@Xiot Thanks for your report!

> However it seems that the prototype isn't actually preserved when the object is used from another module.

It's because we currently do not support that kind of annotation (see https://github.com/dagger/dagger/issues/7401#issuecomment-2125845198)

> First prize, would be to re-attach the prototype chain, but barring that, it would be nice to have an eslint rule to catch things that work slightly differently than a typescript developer would expect.

I'm not sure I understand the idea behind, do you mean having an eslint that can catch unsupported feature of the SDK such as custom inline union type?

**Comment by Xiot** [comment]
Well there two main things here. Well 3 if you count the string literals, but thats for another issue.

The thing that caught me a little off guard was that the prototype chain isn't preserved when you re-hydrate a field.

```
@object()
class RootModule {
  @func() test() {
    return new PModule(new PCheck(4))
  }
}

@object()
class PCheck {
  @field() value: number

  constructor(value: number) {
    this.value = value
  }
  get doubled() {
    return this.value * 2
  }
}

@object()
class PModule {
  @field() value: PCheck
  constructor(value: PCheck) {
    this.value = value
  }
  @func() print() {
    console.log('value', this.value, Object.getPrototypeOf(this.value))
    return this.value.doubled ?? 0
  }
}
```

Calling
```
dagger call test print
```
will return `0` and will print

```
value { value: 4 } [Object: null prototype] {} 
```

That shows that prototype of the `value` field is set to `null`, so the `doubled` property isn't actually there.

My comment about eslint, was coming up with a way to let developers know that even though what they are doing is supported in the language, it won't actually work.

**Comment by Xiot** [comment]
I should have actually created a bug for the prototype not being there, rather than the eslint rule

**Comment by Xiot** [comment]
Created a separate issue for the prototype https://github.com/dagger/dagger/issues/7446

**Comment by TomChv** [comment]
Hey @Xiot, I was thinking about this issue recently, I think I could create some custom linter rules in the TS init template to catch some stuff.

I'm not sure how far I can get this done but here's some ideas I have:

**Checking function**
- raise a warning if the function is exposed but not documented properly

**Checking function argument**
- raise an error if we have an union type because we de not support it for now.
- raise an error if there's a strict `type` as arg, we do not support it for now.
- raise a warning if the args are not documented at the top level comment.
- raise a warning for the casing -> camelcase is expected (I can probably inherit that from another linter)

**Checking function return type**
- Same as 1 & 2 for function argument.
- Assert the func is async if it returns a Promise

**Checking inside the function**
- If a function returns a promise, verify it's correctly awaited, it can avoid a lot of unexpected behaviour, I don't know how to catch that though but that would be great.

**Checking object**
- raise a warning if the object is exposed but not documented propertly.

**Overall code**
- raise an error if the enum isn't correctly formated (non-matching name/values, not readonly...).
- raise an error if a type is decorator -> it cannot be exposed by dagger so it's gonna fail at runtime.

**Comment by Xiot** [comment]
That all sounds good.  
Were you thinking of doing it as an `eslint` rule or something bespoke?

If you do it as a lint rule, I would probably publish it to npm, and then when you init a new typescript project, you can set that up in the newly created project.

**Comment by TomChv** [comment]
Yeah I would like to do it as eslint rules, I've never work on that kind of thing but I'm sure there's nice doc explaining how to do that.

> If you do it as a lint rule, I would probably publish it to npm, and then when you init a new typescript project, you can set that up in the newly created project.

Like uploading it as a npm package so it's downloaded during the init? That would also requiring an eslint setup in the init template

**Comment by Xiot** [comment]
> Yeah I would like to do it as eslint rules, I've never work on that kind of thing but I'm sure there's nice doc explaining how to do that.

The trickest part about doing eslint rules is just understanding the AST, which you already know.

> Like uploading it as a npm package so it's downloaded during the init? 

Well the `yarn install` would pull it down.

> That would also requiring an eslint setup in the init template
You could just make it so that the user has to set it up themselves.  Just have a doc telling people about it.  That way you wouldn't need to do any changes in your set up.
